CONTRIBUTING.md

```diff
@@ -109,48 +109,355 @@ Check out [building instructions](https://github.com/prestodb/presto/tree/master
+                  List<String> columnNames,
+                  RuntimeStats runtimeStats)
+      
+      ```
+
+        * Group lines of logical units together, and use a single empty line to space out the sections.
+
+      ```java
+      public ParquetPageSource(
+                ParquetReader parquetReader,
+                List<Type> types,
+                List<Optional<Field>> fields,
+                List<Boolean> rowIndexLocations,
+                List<String> columnNames,
+                RuntimeStats runtimeStats)
+        {
+            this.parquetReader = requireNonNull(parquetReader, "parquetReader is null");
+            this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+            this.fields = ImmutableList.copyOf(requireNonNull(fields, "fields is null"));
+            this.rowIndexLocations = requireNonNull(rowIndexLocations, "rowIndexLocations is null");
+            this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
+            this.runtimeStats = requireNonNull(runtimeStats, "runtimeStats is null");
+      
+            checkArgument(
+                    types.size() == rowIndexLocations.size() && types.size() == fields.size(),
+                    "types, rowIndexLocations, and fields must correspond one-to-one-to-one");
+      
+            Streams.forEachPair(
+                    rowIndexLocations.stream(),
+                    fields.stream(),
+                    (isIndexColumn, field) -> checkArgument(
+                            !(isIndexColumn && field.isPresent()),
+                            "Field info for row index column must be empty Optional"));
+        }
+      ```
+
+* **Ordering of class members and methods**
+    * Class members are in front of methods.
+    * Group and order the class members and methods in the order of their access levels in descending order:
+        1. public
+        1. protected
+        1. package private
+        1. private
+    * Group and order the class members in descending order:
+        1. static final
+        1. final
+        1. normal
+    * Order the methods with the same access level in the order they’re called.
+* **Encourage proper encapsulation**
+    * Do not use public class members. Use lower access levels as much as possible. Exceptions may apply for performance critical paths, in that case discuss your design first with relevant code owners.
+    * Make a function access level as low as possible
+    * If a public class is only used by one caller and the usage is local, consider making it a nested class.
+* **Immutable and thread safe classes**
+    * Whenever possible, class fields should be final
+    * When it's not possible, ensure that accesses to the non-final fields are thread safe by whatever methods are appropriate for the circumstance (Concurrent collections, synchronized access, etc.) if they are in a code path that is multi-threaded
```

Comment:
Should it be "When it's possible"?

Reply:
I think it's in the context of the bullet above.  When it's not possible for something to be final, we should protect writes to the field.

Reply:
Oh I see